velox/exec/MarkDistinct.cpp

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "MarkDistinct.h"
+#include "velox/common/base/Range.h"
+#include "velox/vector/FlatVector.h"
+
+#include <algorithm>
+#include <utility>
+
+namespace facebook::velox::exec {
+
+MarkDistinct::MarkDistinct(
+    int32_t operatorId,
+    DriverCtx* driverCtx,
+    const std::shared_ptr<const core::MarkDistinctNode>& planNode)
+    : Operator(
+          driverCtx,
+          planNode->outputType(),
+          operatorId,
+          planNode->id(),
+          "MarkDistinct") {
+  auto inputType = planNode->sources()[0]->outputType();
+
+  // Set all input columns as identity projection.
+  for (auto i = 0; i < inputType->size(); ++i) {
+    identityProjections_.emplace_back(i, i);
+  }
+
+  // Set up result projection
+  resultProjections_.emplace_back(0, inputType->size());
+
+  // Initialize groupingset
+  auto numHashers = planNode.get()->getDistinctVariables().size();
+  std::vector<std::unique_ptr<VectorHasher>> hashers;
+  hashers.reserve(numHashers);
+
+  for (const auto& distinctVariable : planNode.get()->getDistinctVariables()) {
+    auto channel = exprToChannel(distinctVariable.get(), inputType);
```

Comment:
Great question! I've added handling as no op because a const value won't affect hash value.

But upon checking using presto, it seems the planner is smart enough to optimize away a const distinct expr (e.g. select sum(distinct 42)). So our handling may be just dead code that never runs in production. Attached a sample plan:

```
 explain analyze WITH sample AS ( SELECT * FROM (VALUES (1,1,1), (1,1,2), (1,1,1), (2,1,2), (2,2,1), (2,3,2)) AS account (c1, c2, c3) ) SELECT c1,sum(distinct 0),sum(distinct c3) from sample group by c1;                                                                               Query Plan                                                                               
------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Fragment 1 [HASH]                                                                                                                                                      
     CPU: 144.69ms, Scheduled: 525.50ms, Input: 6 rows (138B); per task: avg.: 0.20 std.dev.: 0.75, Output: 2 rows (46B)                                                
     Output layout: [field, sum, sum_26]                                                                                                                                
     Output partitioning: SINGLE []                                                                                                                                     
     Stage Execution Strategy: UNGROUPED_EXECUTION                                                                                                                      
     - Aggregate[field] => [field:integer, sum:bigint, sum_26:bigint]                                                                                                   
             CPU: 3.00ms (30.00%), Scheduled: 4.00ms (5.48%), Output: 2 rows (46B)                                                                                      
             Input avg.: 0.01 rows, Input std.dev.: 2188.61%                                                                                                            
             sum := "presto.default.sum"((expr_25)) (mask = expr$distinct) (1:146)                                                                                      
             sum_26 := "presto.default.sum"((c3)) (mask = c3$distinct) (1:162)                                                                                          
         - MarkDistinct[distinct=field:integer, c3:bigint marker=c3$distinct] => [field:integer, expr_25:bigint, c3:bigint, expr$distinct:boolean, c3$distinct:boolean] 
                 CPU: 0.00ns (0.00%), Scheduled: 0.00ns (0.00%), Output: 6 rows (162B)                                                                                  
                 Input avg.: 0.01 rows, Input std.dev.: 2188.61%                                                                                                        
             - MarkDistinct[distinct=field:integer marker=expr$distinct] => [field:integer, expr_25:bigint, c3:bigint, expr$distinct:boolean]                           
                     CPU: 3.00ms (30.00%), Scheduled: 61.00ms (83.56%), Output: 6 rows (150B)                                                                           
                     Input avg.: 0.01 rows, Input std.dev.: 2188.61%                                                                                                    
                 - LocalExchange[HASH] (field) => [field:integer, expr_25:bigint, c3:bigint]                                                                            
                         CPU: 3.00ms (30.00%), Scheduled: 5.00ms (6.85%), Output: 6 rows (138B)                                                                         
                         Input avg.: 0.01 rows, Input std.dev.: 2188.61%                                                                                                
                     - RemoteSource[2] => [field:integer, expr_25:bigint, c3:bigint]                                                                                    
                             CPU: 0.00ns (0.00%), Scheduled: 0.00ns (0.00%), Output: 6 rows (138B)                                                                      
                             Input avg.: 0.01 rows, Input std.dev.: 2188.61%                                                                                            
                                                                                                                                                                        
 Fragment 2 [SINGLE]                                                                                                                                                    
     CPU: 3.63ms, Scheduled: 4.49ms, Input: 6 rows (60B); per task: avg.: 6.00 std.dev.: 0.00, Output: 6 rows (138B)                                                    
     Output layout: [field, expr_25, c3]                                                                                                                                
     Output partitioning: HASH [field, expr_25]                                                                                                                         
     Stage Execution Strategy: UNGROUPED_EXECUTION                                                                                                                      
     - Project[projectLocality = LOCAL] => [field:integer, expr_25:bigint, c3:bigint]                                                                                   
             CPU: 1.00ms (10.00%), Scheduled: 2.00ms (2.74%), Output: 6 rows (138B)                                                                                     
             Input avg.: 0.19 rows, Input std.dev.: 556.78%                                                                                                             
             expr_25 := BIGINT'0'                                                                                                                                       
             c3 := CAST(field_1 AS bigint) (1:36)                                                                                                                       
         - LocalExchange[ROUND_ROBIN] () => [field:integer, field_1:integer]                                                                                            
                 CPU: 0.00ns (0.00%), Scheduled: 1.00ms (1.37%), Output: 6 rows (60B)                                                                                   
                 Input avg.: 6.00 rows, Input std.dev.: 0.00%                                                                                                           
             - Values => [field:integer, field_1:integer]                                                                                                               
                     CPU: 0.00ns (0.00%), Scheduled: 0.00ns (0.00%), Output: 6 rows (60B)                                                                               
                     Input avg.: 6.00 rows, Input std.dev.: 0.00%                                                                                                       
                     (INTEGER'1', INTEGER'1')                                                                                                                           
                     (INTEGER'1', INTEGER'2')                                                                                                                           
                     (INTEGER'1', INTEGER'1')                                                                                                                           
                     (INTEGER'2', INTEGER'2')                                                                                                                           
                     (INTEGER'2', INTEGER'1')                                                                                                                           
                     (INTEGER'2', INTEGER'2')                                                                                                                           
                                                                                                                                                                        
                                                                                                                                                                        
(1 row)
```